crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/meta_net.rs

```diff
@@ -824,23 +877,44 @@ impl MetaNet {
             preflight: Some((
                 Arc::new(move |_| {
                     let i_s = kitsune_internal_sender.clone();
+                    let user_data_sent = user_data_sent.clone();
 
                     Box::pin(async move {
                         let agent_list = i_s
                             .get_all_local_joined_agent_infos()
                             .await
                             .unwrap_or_default();
-                        wire::Wire::peer_unsolicited(agent_list).encode_vec()
+                        PreflightData::v0(KITSUNE_PROTOCOL_VERSION, agent_list, user_data_sent)
+                            .encode_vec()
                     })
                 }),
-                Arc::new(move |_, data| {
+                Arc::new(move |url, data| {
                     let e_s = evt_sender.clone();
-                    Box::pin(async move {
-                        match wire::Wire::decode_ref(&data) {
-                            Ok((
-                                _,
-                                wire::Wire::PeerUnsolicited(wire::PeerUnsolicited { peer_list }),
-                            )) => {
+                    let url = url.clone();
+                    match PreflightData::decode_ref(&data) {
+                        Ok((
+                            _,
+                            PreflightData::V0(V0 {
+                                kitsune_protocol_version,
+                                peer_list,
+                                user_data: user_data_bytes_received,
+                            }),
+                        )) => {
+                            if kitsune_protocol_version != KITSUNE_PROTOCOL_VERSION {
+                                tracing::warn!(
+                                    ?url,
+                                    "kitsune protocol version mismatch: ours = {}, theirs = {}",
+                                    KITSUNE_PROTOCOL_VERSION,
+                                    kitsune_protocol_version,
+                                );
+                                return box_fut_plain(Ok(()));
+                            }
+
+                            if let Err(reason) = user_data_cmp(&url, &user_data_bytes_received) {
+                                tracing::warn!(?url, "tx5 preflight user_data mismatch");
+                                return box_fut_plain(Ok(()));
+                            }
```

Comment:
I don't think it's correct that these return `Ok(())` here... because the connection will be allowed to proceed, but we won't have processed the peer list because they return before we fall through that code below.

Reply:
I actually don't understand why we're allowing the connection to continue below if there is an error either... don't we want to close it if it's incompatible?

Reply:
Oh I'm confused then. I thought avoiding processing the peer list is all we wanted to do, i.e. that's what prevents the connection from being established. The other error cases just return Ok without processing peers.

So what should we actually do on mismatch? Return Err?

Reply:
> I actually don't understand why we're allowing the connection to continue below if there is an error either...

Yeah I was wondering that too, that's how it was set up before

Reply:
I think maybe we were being overly cautious by just putting a warning but letting the connection try to continue. I think, instead, this is a place where we should fail fast and fix whatever problems arise.

Reply:
Yes, returning an `Err` will close the connection.